ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -0,0 +1,888 @@
+    if (lastSegmentEndOffset.compareTo(srcIndex.journal.getFirstOffset()) >= 0) {
+      throw new IllegalArgumentException("Some of the offsets provided for compaction are within the journal");
+    }
+  }
+
+  /**
+   * Copies valid data from the source log segments into the swap spaces.
+   * <p/>
+   * Splits the compaction into two cycles if there aren't enough swap spaces and completes the copy for the current
+   * cycle.
+   * @throws InterruptedException if the compaction was interrupted
+   * @throws IOException if there were I/O errors during copying.
+   * @throws StoreException if there were exceptions reading to writing to store components.
+   */
+  private void copy() throws InterruptedException, IOException, StoreException {
+    setupState();
+    List<String> logSegmentsUnderCompaction = compactionLog.getCompactionDetails().getLogSegmentsUnderCompaction();
+    Offset startOffsetOfLastIndexSegmentForDelete = getStartOffsetOfLastIndexSegmentForDelete();
+    FileSpan duplicateSearchSpan = null;
+    if (compactionLog.getCurrentIdx() > 0) {
+      // only records in the the very first log segment in the cycle can have been copied in a previous cycle
+      LogSegment firstLogSegment = srcLog.getSegment(logSegmentsUnderCompaction.get(0));
+      LogSegment prevSegment = srcLog.getPrevSegment(firstLogSegment);
+      if (prevSegment != null) {
+        // duplicate data, if it exists, can only be in the log segment just before the first log segment in the cycle.
+        Offset startOffset = new Offset(prevSegment.getName(), prevSegment.getStartOffset());
+        Offset endOffset = new Offset(prevSegment.getName(), prevSegment.getEndOffset());
+        duplicateSearchSpan = new FileSpan(startOffset, endOffset);
+      }
+    }
+
+    for (String logSegmentName : logSegmentsUnderCompaction) {
+      LogSegment srcLogSegment = srcLog.getSegment(logSegmentName);
+      Offset logSegmentEndOffset = new Offset(srcLogSegment.getName(), srcLogSegment.getEndOffset());
+      if (needsCopying(logSegmentEndOffset) && !copyDataByLogSegment(srcLogSegment, duplicateSearchSpan,
+          startOffsetOfLastIndexSegmentForDelete)) {
+        if (isActive) {
+          // split the cycle only if there is no shutdown in progress
+          compactionLog.splitCurrentCycle(logSegmentName);
+        }
+        break;
+      }
+      // only records in the the very first log segment in the cycle can have been copied in a previous cycle
+      duplicateSearchSpan = null;
+    }
+
+    numSwapsUsed = tgtIndex.getLogSegmentCount();
+    tgtIndex.close();
+    tgtLog.close();
+    // persist the bloom of the "latest" index segment if it exists
+    if (numSwapsUsed > 0) {
+      tgtIndex.getIndexSegments().lastEntry().getValue().map(true);
+    } else {
+      // there were no valid entries copied, return any temp segments back to the pool
+      File[] files = dataDir.listFiles(TEMP_LOG_SEGMENTS_FILTER);
```

Comment:
Is this "giving back" better as a method of its own? Also, at this point is it not good to commit even if the file deletion fails or any I/O error happens?